File: cses/04_graph_algorithms/1691_mail-delivery/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    m = int(data[idx]); idx += 1

    # Đồ thị vô hướng: adj[v] chứa các cặp (đỉnh kề, edge id).
    adj = [[] for _ in range(n + 1)]
    deg = [0] * (n + 1)

    for e in range(m):
        a = int(data[idx]); idx += 1
        b = int(data[idx]); idx += 1
        adj[a].append((b, e))
        adj[b].append((a, e))
        deg[a] += 1
        deg[b] += 1

    out = sys.stdout

    # Điều kiện cần của chu trình Euler: mọi đỉnh phải có bậc chẵn.
    for v in range(1, n + 1):
        if deg[v] & 1:
            out.write("IMPOSSIBLE\n")
            return

    used = bytearray(m)          # đánh dấu cạnh đã đi qua
    ptr = [0] * (n + 1)          # con trỏ duyệt adjacency của mỗi đỉnh

    # Thuật toán Hierholzer (iterative để tránh tràn đệ quy): đi theo cạnh chưa dùng
    # cho tới khi kẹt, rồi lùi lại và ghi đỉnh vào circuit.
    stack = [1]
    circuit = []
    while stack:
        v = stack[-1]
        av = adj[v]
        p = ptr[v]
        lv = len(av)
        # Bỏ qua các cạnh đã dùng nhờ con trỏ ptr (mỗi cạnh chỉ bị bỏ qua O(1) lần).
        while p < lv and used[av[p][1]]:
            p += 1
        ptr[v] = p
        if p == lv:
            # Không còn cạnh: chốt đỉnh này vào circuit và lùi lại.
            circuit.append(v)
            stack.pop()
        else:
            # Đi tiếp qua một cạnh chưa dùng, đánh dấu cạnh và tiến sang đỉnh kề.
            to, eid = av[p]
            used[eid] = 1
            ptr[v] = p + 1
            stack.append(to)

    # Chu trình hợp lệ phải đi qua tất cả m cạnh -> có m + 1 đỉnh. Nếu ít hơn nghĩa là
    # đồ thị không liên thông (còn cạnh ở thành phần khác) -> in "IMPOSSIBLE".
    if len(circuit) != m + 1:
        out.write("IMPOSSIBLE\n")
        return

    # circuit đang theo thứ tự ngược, đảo lại để có lộ trình đi đúng chiều.
    circuit.reverse()
    out.write(" ".join(map(str, circuit)))
    out.write("\n")


main()

File: cses/04_graph_algorithms/1691_mail-delivery/solution.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

// Một phần tử trong adjacency list: đỉnh kề và id của cạnh (dùng chung cho cả hai chiều).
struct AdjacentEdge {
    int destination;
    int id;
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    // Đọc đồ thị vô hướng: mỗi cạnh id được thêm vào adjacency của cả hai đầu mút.
    vector<vector<AdjacentEdge>> graph(n + 1);
    vector<int> degree(n + 1, 0);
    for (int id = 0; id < m; ++id) {
        int first, second;
        cin >> first >> second;
        graph[first].push_back({second, id});
        graph[second].push_back({first, id});
        ++degree[first];
        ++degree[second];
    }

    // Điều kiện cần của chu trình Euler: mọi đỉnh phải có bậc chẵn.
    for (int vertex = 1; vertex <= n; ++vertex) {
        if (degree[vertex] % 2 != 0) {
            cout << "IMPOSSIBLE\n";
            return 0;
        }
    }

    vector<char> used(m, false);        // đánh dấu cạnh đã đi qua
    vector<size_t> nextEdge(n + 1, 0);  // con trỏ duyệt adjacency của mỗi đỉnh
    vector<int> stack{1};               // Hierholzer iterative, bắt đầu từ đỉnh 1
    vector<int> circuit;
    circuit.reserve(m + 1);

    // Thuật toán Hierholzer: đi theo cạnh chưa dùng cho tới khi kẹt, rồi lùi (backtrack)
    // và ghi đỉnh vào circuit.
    while (!stack.empty()) {
        const int vertex = stack.back();
        // Bỏ qua các cạnh đã đi qua nhờ con trỏ nextEdge (mỗi cạnh chỉ bị bỏ qua O(1) lần).
        while (nextEdge[vertex] < graph[vertex].size() &&
               used[graph[vertex][nextEdge[vertex]].id]) {
            ++nextEdge[vertex];
        }

        if (nextEdge[vertex] == graph[vertex].size()) {
            // Không còn cạnh nào: chốt đỉnh này vào circuit và lùi lại.
            circuit.push_back(vertex);
            stack.pop_back();
        } else {
            // Đi tiếp qua một cạnh chưa dùng, đánh dấu cạnh và tiến sang đỉnh kề.
            const AdjacentEdge edge = graph[vertex][nextEdge[vertex]++];
            used[edge.id] = true;
            stack.push_back(edge.destination);
        }
    }

    // Chu trình hợp lệ phải đi qua tất cả m cạnh -> có đúng m + 1 đỉnh.
    // Nếu thiếu nghĩa là đồ thị không liên thông (còn cạnh ở thành phần khác).
    if (static_cast<int>(circuit.size()) != m + 1) {
        cout << "IMPOSSIBLE\n";
        return 0;
    }

    // circuit đang theo thứ tự ngược, đảo lại để có lộ trình đi đúng chiều.
    reverse(circuit.begin(), circuit.end());
    for (size_t i = 0; i < circuit.size(); ++i) {
        if (i > 0) {
            cout << ' ';
        }
        cout << circuit[i];
    }
    cout << '\n';
    return 0;
}
